lib/ASTGen/Sources/ASTGen/ASTGen.swift

```diff
@@ -150,29 +148,122 @@ extension ASTGenVisitor {
 // Misc visits.
 // TODO: Some of these are called within a single file/method; we may want to move them to the respective files.
 extension ASTGenVisitor {
-  public func visit(_ node: MemberBlockItemSyntax) -> ASTNode {
+  public func generate(_ node: MemberBlockItemSyntax) -> ASTNode {
     visit(Syntax(node.decl))
   }
 
-  public func visit(_ node: InitializerClauseSyntax) -> ASTNode {
-    visit(node.value)
+  public func generate(_ node: InitializerClauseSyntax) -> ASTNode {
+    visit(Syntax(node.value))
   }
 
-  public func visit(_ node: ConditionElementSyntax) -> ASTNode {
-    visit(node.condition)
+  public func generate(_ node: ConditionElementSyntax) -> ASTNode {
+    visit(Syntax(node.condition))
   }
 
-  public func visit(_ node: CodeBlockItemSyntax) -> ASTNode {
-    visit(node.item)
+  public func generate(_ node: CodeBlockItemSyntax) -> ASTNode {
+    visit(Syntax(node.item))
   }
 
-  public func visit(_ node: ArrayElementSyntax) -> ASTNode {
-    visit(node.expression)
+  public func generate(_ node: ArrayElementSyntax) -> ASTNode {
+    visit(Syntax(node.expression))
```

Comment:
You could add the following function. That way you don’t need to perform these upcasts to `Syntax`.

```swift
func generate(_ node: ExprSyntax) {
  return generate(Syntax(node))
}
```

Similar for all the other base nodes (`DeclSyntax`, `ExprSyntax`, `PatternSyntax`, `StmtSyntax`, `TypeSyntax`)

Reply:
How do I find the base nodes for the nodes that are being passed? What would be the base node for `ConditionElementSyntax` ? 

I found this when looking after jumping to the definition

```swift
// MARK: - ConditionElementSyntax

/// ### Children
/// 
///  - `condition`: (``ExprSyntax`` | ``AvailabilityConditionSyntax`` | ``MatchingPatternConditionSyntax`` | ``OptionalBindingConditionSyntax``)
///  - `trailingComma`: `','`?
///
/// ### Contained in
/// 
///  - ``ConditionElementListSyntax``
public struct ConditionElementSyntax: SyntaxProtocol, SyntaxHashable, _LeafSyntaxNodeProtocol {
  public enum Condition: SyntaxChildChoices, SyntaxHashable {
    case `expression`(ExprSyntax)
    case `availability`(AvailabilityConditionSyntax)
    case `matchingPattern`(MatchingPatternConditionSyntax)
    case `optionalBinding`(OptionalBindingConditionSyntax)
    ```
Is it based off the children?

Reply:
Would adding the function solve issues with upcasts to `Syntax` elsewhere also? Something on line 83 in `Exprs.swift`.

``` swift
    let expressions = node.lazy.map {
      self.generate(Syntax($0.expression)).rawValue
    }
```
Or do I need to keep the upcast?

Reply:
> How do I find the base nodes for the nodes that are being passed? What would be the base node for `ConditionElementSyntax`?

The base node of `ConditionElementSyntax` is `Syntax`. `Syntax` is basically the type-erased wrapper of `SyntaxProtocol`, which `ConditionElementSyntax` conforms to. You can view all the base nodes in SyntaxBaseNodes.swift.

> Or do I need to keep the upcast?

If it compiles without, you should be able to remove it.